crates/cext/src/sparse_observable.rs

```diff
@@ -0,0 +1,749 @@
+    bit_terms: *mut BitTerm,
+    indices: *mut u32,
+    boundaries: *mut usize,
+) -> *mut SparseObservable {
+    let num_terms = num_terms as usize;
+    let num_bits = num_bits as usize;
+
+    let coeffs = unsafe { ::std::slice::from_raw_parts(coeffs, num_terms).to_vec() };
+    let bit_terms = unsafe { ::std::slice::from_raw_parts(bit_terms, num_bits).to_vec() };
+    let indices = unsafe { ::std::slice::from_raw_parts(indices, num_bits).to_vec() };
+    let boundaries = unsafe { ::std::slice::from_raw_parts(boundaries, num_terms + 1).to_vec() };
+
+    let result = SparseObservable::new(num_qubits, coeffs, bit_terms, indices, boundaries);
+    match result {
+        Ok(obs) => Box::into_raw(Box::new(obs)),
+        Err(_) => ::std::ptr::null_mut(),
+    }
+}
+
+/// @ingroup QkSparseObservable
+/// Free the observable.
+///
+/// @param obs A pointer to the observable to free.
+///
+/// # Example
+///
+///     QkSparseObservable *obs = qk_obs_zero(100);
+///     qk_obs_free(obs);
+///
+/// # Safety
+///
+/// This is unsafe as it is reading and freeing raw memory, which could lead to problems
+/// e.g. if the same pointer is freed multiple times.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_free(obs: *mut SparseObservable) {
+    if !obs.is_null() {
+        unsafe {
+            let _ = Box::from_raw(obs);
+        }
+    }
+}
+
+/// @ingroup QkSparseObservable
+/// Add a term to the observable.
+///
+/// @param obs A pointer to the observable.
+/// @param A pointer to the term to add.
+///
+/// @return An exit code. This is ``>0`` if the term is incoherent or adding the term fails.
+///
+/// # Example
+///
+///     uint32_t num_qubits = 100;
+///     QkSparseObservable *obs = qk_obs_zero(num_qubits);
+///
+///     complex double coeff = 1;
+///     QkBitTerm bit_terms[3] = {QkBitTerm_X, QkBitTerm_Y, QkBitTerm_Z};
+///     uint32_t indices[3] = {0, 1, 2};
+///     QkSparseTerm term = {&coeff, 3, bit_terms, indices, num_qubits};
+///
+///     int exit_code = qk_obs_add_term(obs, &term);
+///
+/// # Safety
+///
+/// Behavior is undefined if any of the following is violated:
+///
+///   * ``obs`` is a valid, non-null pointer to a ``QkSparseObservable``
+///   * ``cterm`` is a valid, non-null pointer to a ``QkSparseTerm``
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_add_term(
+    obs: *mut SparseObservable,
+    cterm: *const CSparseTerm,
+) -> ExitCode {
+    // check pointers are valid and cast them to a reference
+    let obs = unsafe { mut_ptr_as_ref(obs) };
+    let cterm = unsafe { const_ptr_as_ref(cterm) };
+
+    let view = match cterm.try_into() {
+        Ok(view) => view,
+        Err(err) => return ExitCode::from(err),
+    };
+
+    match obs.add_term(view) {
+        Ok(_) => ExitCode::Success,
+        Err(err) => ExitCode::from(err),
+    }
+}
+
+/// @ingroup QkSparseObservable
+/// Get an observable term by reference. This can modify the underlying observable.
```

Comment:
Should maybe comment that it's typically a _really bad_ idea to actually modify it in place, or at least spell out all the reasons it's hard (e.g. can't change the length of indices or terms, the `coeff` won't update (subject to my other comment about `*mut Complex64` maybe not being necessary), etc) - the fact that you _can_ mutate the observable in place through these pointers is more of a side-effect of being no-copy than a desirable feature.